[optim] Make casting to match params a hook (2nd try)

janeyx99 opened the pull request:
Reland of #106725, but instead of registering the hook in the constructor, do it lazily through @property

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108303

janeyx99 committed:
[optim] Make casting to match params a hook (2nd try)

[ghstack-poisoned]

Skylion007 commented on torch/optim/optimizer.py:
```diff
@@ -233,17 +233,73 @@ class Optimizer:
+        """
+
+        if len(groups) != len(saved_groups):
+            raise ValueError("loaded state dict has a different number of "
+                             "parameter groups")
```
Could you add the relevant length to the error message to prevent someone going crazy later :P

Skylion007 commented on torch/optim/optimizer.py:
```diff
@@ -233,17 +233,73 @@ class Optimizer:
+                return value
+
+        # Cast state tensors to appropriate types in the state_dict
+        for k in state_dict['state']:
+            v = state_dict['state'][k]
```
```suggestion
        for k, v in state_dict['state'].items():
```

janeyx99 committed:
Update on "[optim] Make casting to match params a hook (2nd try)"

Reland of #106725, but instead of registering the hook in the constructor, do it lazily through property




[ghstack-poisoned]

albanD commented on torch/optim/optimizer.py:
```diff
@@ -650,6 +754,14 @@ def register_load_state_dict_pre_hook(
 
+        .. note::
+
+            There is an automatically registered load_state_dict pre-hook which
```
Is it correct that because we always access the attribute before registering the user hook, the lazily added one will always be registered before any other hook?

albanD commented on torch/optim/optimizer.py:
```diff
@@ -403,6 +463,50 @@ def _patch_step_function(self) -> None:
+        self._set_up_load_state_dict_pre_hooks_and_handle()
+        return self._load_state_dict_cast_handle_legit
+
+    @_load_state_dict_cast_handle.setter
```
Why have this? You never set it?

albanD commented on torch/optim/optimizer.py:
```diff
@@ -233,17 +233,73 @@ class Optimizer:
         self._optimizer_state_dict_pre_hooks = OrderedDict()
         self._optimizer_state_dict_post_hooks = OrderedDict()
-        self._optimizer_load_state_dict_pre_hooks = OrderedDict()
+        # _optimizer_load_state_dict_pre_hooks and _load_state_dict_cast_handle
```
Comments explaining why they are not here is good. Please make all 3 consistent though

albanD commented on torch/optim/optimizer.py:
```diff
@@ -403,6 +463,50 @@ def _patch_step_function(self) -> None:
+        self._optimizer_load_state_dict_pre_hooks_legit[handle.id] = Optimizer._cast_state_to_match_params_hook
+
+    @property
+    def _optimizer_load_state_dict_pre_hooks(
```
Why do you need to make this one lazy as well? Is it because you use that as a way to trigger the lazy initialization of the handle and this is more reliable to do it here rather than in the register_hook method?
If so, a small comment here explaining that would be good.

albanD commented on torch/optim/optimizer.py:
```diff
@@ -403,6 +463,50 @@ def _patch_step_function(self) -> None:
+        # otherwise, set up the hooks and handle
+        if (hasattr(self, "_load_state_dict_cast_handle_legit") and
+                self._load_state_dict_cast_handle_legit is not None):
+            raise RuntimeError("the load_state_dict pre hooks should always be init'd "
```
Is that going to play well with serialization? Are we sure that there is no way we can have one but not the other? (both when users do the right thing of using state_dict and the wrong thing of serializing the whole Optimizer)

janeyx99 commented:
not stale!
